# Risk Rule Testing & Validation Guide

**Purpose**: Validate that all 13 risk rules have correct arithmetic, breach detection, and enforcement

**Date**: 2025-10-28
**Status**: Testing framework complete, ready for individual rule validation

---

## üéØ What We're Testing

For each of the 13 rules, we need to validate 3 things:

1. **Arithmetic** - Calculations are mathematically correct
2. **Breach Detection** - Violations trigger at exact thresholds
3. **Enforcement** - SDK actions execute correctly (close positions, reject orders, etc.)

---

## üîç Event Flow Architecture

### Current System (After Latest Fixes)

```
TopstepX API (Live Market)
    ‚Üì
SignalR WebSocket Callbacks
    ‚îú‚îÄ position_update ‚Üí POSITION_UPDATED event
    ‚îú‚îÄ order_update ‚Üí ORDER_PLACED/FILLED/CANCELLED events
    ‚îú‚îÄ trade_update ‚Üí TRADE_EXECUTED event
    ‚îú‚îÄ account_update ‚Üí Account balance updates
    ‚îî‚îÄ quote_update ‚Üí MARKET_DATA_UPDATED event ‚úÖ NEW!
    ‚Üì
EventBus.publish(RiskEvent)
    ‚Üì
engine.evaluate_rules(event)
    ‚îú‚îÄ rule1.evaluate()
    ‚îú‚îÄ rule2.evaluate()
    ‚îî‚îÄ rule3.evaluate()
    ‚Üì
If violation detected:
    ‚Üì
engine._handle_violation()
    ‚îú‚îÄ action="flatten" ‚Üí trading_integration.flatten_all()
    ‚îú‚îÄ action="close_position" ‚Üí trading_integration.flatten_position(symbol)
    ‚îú‚îÄ action="pause" ‚Üí trading_integration.pause_trading()
    ‚îî‚îÄ action="alert" ‚Üí Log/notify only
    ‚Üì
SDK Enforcement
    ‚îî‚îÄ suite[symbol].positions.close_all_positions()
    ‚îî‚îÄ suite[symbol].orders.place_market_order()
```

### Key Changes Made Today

‚úÖ **Fixed**: Removed `"orderbook"` from SDK features (was causing depth entry errors)
‚úÖ **Added**: Quote subscription via `quote_update` callback for real-time prices
‚úÖ **Added**: `EventType.MARKET_DATA_UPDATED` for market price events
‚úÖ **Created**: `test_rule_validation.py` framework for testing all rules

---

## üß™ Testing Methods

### Method 1: Mock Event Injection (Fast, Isolated)

**Use Case**: Test arithmetic and breach detection without live SDK

```bash
# Test all rules
python test_rule_validation.py

# Test specific rule
python test_rule_validation.py --rule RULE-003

# Verbose output
python test_rule_validation.py --verbose
```

**How It Works**:
1. Creates in-memory database (no persistence)
2. Instantiates rule with test config
3. Injects mock RiskEvents with specific values
4. Validates that violations trigger at correct thresholds
5. Checks enforcement actions are correct

**Example: RULE-003 (Daily Realized Loss)**
```python
# Configure: limit = -$500
# Inject: Trade 1 (-$200) ‚Üí No violation
# Inject: Trade 2 (-$150) ‚Üí No violation (total: -$350)
# Inject: Trade 3 (-$200) ‚Üí VIOLATION! (total: -$550 > -$500)

# Validate:
assert len(violations) == 1  # Only trade 3 violates
assert violations[0]["action"] == "flatten"
assert pnl_tracker.get_daily_pnl() == -550.0
```

### Method 2: Live SDK Testing (Real Market Data)

**Use Case**: Validate end-to-end with live TopstepX connection

```bash
# Run development runtime
python run_dev.py

# Watch for real events and rule evaluations
# Look for 8-checkpoint logs to verify flow
```

**What To Watch For**:
1. ‚úÖ Checkpoint 6: Event received (position/order/trade/quote)
2. ‚úÖ Checkpoint 7: Rule evaluated (PASSED or VIOLATED)
3. ‚úÖ Checkpoint 8: Enforcement triggered (if violation)
4. ‚úÖ SDK enforcement executes (positions closed, orders cancelled)

**Manual Test Scenarios**:

For each rule, you need to manually trigger the violation condition:

| Rule | Manual Test Scenario |
|------|----------------------|
| RULE-001 | Open 6 contracts (limit: 5) ‚Üí Should close 1 |
| RULE-002 | Open 3 MNQ (limit: 2 per symbol) ‚Üí Should close 1 MNQ |
| RULE-003 | Lose $550 in trades (limit: -$500) ‚Üí Should flatten all |
| RULE-004 | Let position float to -$750 unrealized (limit: -$750) ‚Üí Should close |
| RULE-005 | Let position hit +$500 profit (target: $500) ‚Üí Should take profit |
| RULE-006 | Execute 4 trades in 1 minute (limit: 3/min) ‚Üí Should reject 4th |
| RULE-007 | Lose $150 in one trade (threshold: -$100) ‚Üí Should cooldown |
| RULE-008 | Open position without stop-loss ‚Üí Should reject (if enabled) |
| RULE-009 | Try to trade outside 8:30-15:00 CT ‚Üí Should reject |
| RULE-011 | Try to trade blocked symbol (e.g., ES) ‚Üí Should reject |
| RULE-012 | Open position ‚Üí Should auto-place bracket orders |
| RULE-013 | Make $1000 profit (target: $1000) ‚Üí Should stop trading |

### Method 3: E2E Tests (Automated)

**Location**: `tests/e2e/`

**Status**: 35/37 E2E tests passing (95%)

```bash
# Run E2E tests
python run_tests.py ‚Üí [4] E2E tests

# These tests use real SDK but mocked positions/orders
# They validate the complete flow without requiring manual trading
```

---

## üìã Rule-by-Rule Testing Checklist

### Trade-by-Trade Rules (6 rules)

#### ‚úÖ RULE-001: Max Contracts
- [ ] Test arithmetic: 2 + 1 + 3 = 6 > 5 (violates)
- [ ] Test enforcement: `close_position` action triggers
- [ ] Test SDK call: `suite[symbol].positions.close_all_positions()`
- [ ] Verify NO lockout (can trade again immediately)

#### ‚úÖ RULE-002: Max Contracts Per Instrument
- [ ] Test per-symbol limits: MNQ=2, ES=1, NQ=unlimited
- [ ] Test arithmetic: 3 MNQ > 2 (violates), but 1 ES is OK
- [ ] Test enforcement: Only close excess MNQ contracts
- [ ] Verify other symbols unaffected

#### ‚è∏Ô∏è RULE-004: Daily Unrealized Loss
- [ ] Test real-time price updates (requires `MARKET_DATA_UPDATED` events)
- [ ] Test arithmetic: entry_price=21500, current_price=21400, size=5
  - Unrealized P&L = (21400 - 21500) * 5 * tick_value = loss
- [ ] Test enforcement: Close position when unrealized loss hits limit
- [ ] Test quote subscription working (see quote_update callback)

#### ‚è∏Ô∏è RULE-005: Max Unrealized Profit
- [ ] Test profit target calculation
- [ ] Test enforcement: Take profit at target
- [ ] Test partial close option (if configured)

#### ‚è∏Ô∏è RULE-008: No Stop-Loss Grace
- [ ] Test rejection when position opened without stop
- [ ] Test grace period (60s to place stop)
- [ ] Test enforcement after grace expires

#### ‚è∏Ô∏è RULE-011: Symbol Blocks
- [ ] Test symbol blocking (ES, NQ)
- [ ] Test wildcard patterns (ES*, *USD)
- [ ] Test enforcement: reject_order or close_position

#### ‚è∏Ô∏è RULE-012: Trade Management
- [ ] Test auto-bracket placement
- [ ] Test breakeven adjustment (after X ticks profit)
- [ ] Test trailing stop (after Y ticks profit)

### Timer/Cooldown Rules (2 rules)

#### ‚è∏Ô∏è RULE-006: Trade Frequency Limit
- [ ] Test per_minute limit (3 trades/min)
- [ ] Test per_hour limit (10 trades/hr)
- [ ] Test per_session limit (50 trades/day)
- [ ] Test enforcement: reject_order during cooldown
- [ ] Test auto-unlock after timer expires

#### ‚è∏Ô∏è RULE-007: Cooldown After Loss
- [ ] Test tiered cooldowns:
  - -$100 loss ‚Üí 5 min cooldown
  - -$200 loss ‚Üí 15 min cooldown
  - -$300 loss ‚Üí 30 min cooldown
- [ ] Test enforcement: flatten + lockout
- [ ] Test timer persistence (crash recovery)

### Hard Lockout Rules (4 rules)

#### ‚úÖ RULE-003: Daily Realized Loss
- [ ] Test arithmetic: cumulative daily P&L tracking
- [ ] Test enforcement: flatten + hard lockout
- [ ] Test lockout persists until 5 PM ET reset
- [ ] Test database persistence (survives restart)

#### ‚è∏Ô∏è RULE-013: Daily Realized Profit
- [ ] Test profit target: $1000
- [ ] Test enforcement: flatten + hard lockout
- [ ] Test message: "Good job! See you tomorrow."
- [ ] Test reset at 5 PM ET

#### ‚è∏Ô∏è RULE-009: Session Block Outside
- [ ] Test time range: 8:30 AM - 3:00 PM CT
- [ ] Test timezone handling (DST aware)
- [ ] Test enforcement: reject_order outside hours
- [ ] Test holiday calendar (if enabled)

#### ‚è∏Ô∏è RULE-010: Auth Loss Guard
- [ ] Test connection loss detection
- [ ] Test `canTrade=false` detection
- [ ] Test alert-only (no enforcement)
- [ ] Test reconnection handling

---

## üöÄ Quick Start: Testing Your First Rule

Let's test RULE-003 (Daily Realized Loss) as an example:

### Step 1: Run Mock Test

```bash
python test_rule_validation.py --rule RULE-003 --verbose
```

**Expected Output**:
```
Testing RULE-003: Daily Realized Loss
Trade 1: -$200 loss
  Current P&L: $-200.00
  ‚úÖ No violation
Trade 2: -$150 loss
  Current P&L: $-350.00
  ‚úÖ No violation
Trade 3: -$200 loss
  Current P&L: $-550.00
  ‚ùå VIOLATION: Daily loss limit breached
  Enforcement action: flatten

‚úÖ RULE-003 PASSED
```

### Step 2: Run Live Test

```bash
# Start development runtime
python run_dev.py

# In another terminal, monitor logs
tail -f data/logs/risk_manager.log | grep "RULE-003"
```

### Step 3: Manually Trigger Violation

1. Place trades that lose money
2. Watch for rule evaluation logs
3. When total loss > -$500, rule should trigger
4. Verify enforcement: All positions flattened
5. Verify lockout: Can't trade until 5 PM ET

---

## üìä P&L Calculation Integration

### The Missing Link (NOW FIXED)

**Before Today**:
- ‚ùå No quote subscription ‚Üí No real-time prices
- ‚ùå Unrealized PnL couldn't be calculated
- ‚ùå RULE-004 and RULE-005 wouldn't work

**After Today**:
- ‚úÖ Quote subscription added (`_on_quote_update`)
- ‚úÖ `MARKET_DATA_UPDATED` events published
- ‚úÖ Real-time prices available in `engine.market_prices`
- ‚úÖ Rules can calculate unrealized P&L

### How PnL Calculation Works Now

```python
# 1. Quote arrives from SDK
def _on_quote_update(self, data):
    # Extract price
    symbol = data["symbol"]
    last_price = data["last"]

    # Publish market data event
    event = RiskEvent(
        event_type=EventType.MARKET_DATA_UPDATED,
        data={"symbol": symbol, "price": last_price}
    )
    await self.event_bus.publish(event)

# 2. Rule receives event and calculates P&L
async def evaluate(self, event: RiskEvent, engine: RiskEngine):
    if event.event_type == EventType.MARKET_DATA_UPDATED:
        symbol = event.data["symbol"]
        current_price = event.data["price"]

        # Get position
        position = engine.current_positions.get(symbol)
        if position:
            entry_price = position["average_price"]
            size = position["size"]
            tick_value = TICK_VALUES[symbol]

            # Calculate unrealized P&L
            unrealized_pnl = (current_price - entry_price) * size * tick_value

            # Check against limit
            if unrealized_pnl <= self.limit:  # e.g., -800 <= -750
                return {
                    "action": "close_position",
                    "symbol": symbol,
                    "unrealized_pnl": unrealized_pnl
                }
```

---

## üêõ Troubleshooting

### Issue: "Depth entry error"
**Status**: ‚úÖ FIXED
**Solution**: Removed `"orderbook"` from SDK features

### Issue: "No quote updates"
**Status**: ‚úÖ FIXED
**Solution**: Added `quote_update` callback

### Issue: "Rule doesn't trigger"
**Possible Causes**:
1. Event type mismatch (check `rule.evaluate()` event filter)
2. Arithmetic wrong (add debug logs to see calculations)
3. Threshold not reached (verify test values)
4. PnL tracker not updated (must call `add_trade_pnl()`)

### Issue: "Enforcement doesn't execute"
**Possible Causes**:
1. `trading_integration` not wired to engine
2. SDK not connected
3. Action name mismatch (check spelling: "flatten" not "flatten_all")
4. Exception in enforcement code (check logs)

---

## üìà Progress Tracking

| Rule | Mock Test | Live Test | E2E Test | Status |
|------|-----------|-----------|----------|--------|
| RULE-001 | ‚úÖ | ‚è∏Ô∏è | ‚úÖ | Ready |
| RULE-002 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |
| RULE-003 | ‚úÖ | ‚è∏Ô∏è | ‚úÖ | Ready |
| RULE-004 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs quote test |
| RULE-005 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs quote test |
| RULE-006 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |
| RULE-007 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |
| RULE-008 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |
| RULE-009 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |
| RULE-010 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚è∏Ô∏è | Alert-only |
| RULE-011 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |
| RULE-012 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |
| RULE-013 | ‚è∏Ô∏è | ‚è∏Ô∏è | ‚úÖ | Needs mock test |

---

## üéØ Next Steps

### Immediate (Today):
1. ‚úÖ Fix orderbook depth error
2. ‚úÖ Add quote subscription
3. ‚úÖ Create testing framework
4. ‚è∏Ô∏è Run `python test_rule_validation.py` to verify RULE-001 and RULE-003
5. ‚è∏Ô∏è Add remaining 11 rule tests to framework

### Short-term (This Week):
1. Complete all 13 mock tests in `test_rule_validation.py`
2. Run live tests for critical rules (RULE-003, RULE-001, RULE-006)
3. Verify quote subscription works with real market data
4. Test unrealized PnL calculations with live prices

### Long-term (Before Production):
1. Complete all live manual tests
2. Verify enforcement actions execute correctly
3. Test crash recovery (lockout/timer persistence)
4. Performance test: High-frequency quote updates
5. Integration test: Multiple rules interacting

---

## üìù Notes

### Design Decisions

**Why mock tests first?**
- Fast feedback loop (no SDK delays)
- Isolated testing (one rule at a time)
- Arithmetic validation (exact threshold testing)
- No live trading risk

**Why live tests required?**
- Validate SDK integration actually works
- Test quote subscription with real market data
- Verify enforcement actions execute correctly
- Test edge cases (connection loss, API errors)

**Why E2E tests matter?**
- Automated regression testing
- CI/CD integration
- Comprehensive scenario coverage
- No manual effort required

### Testing Philosophy

1. **Mock First**: Validate logic and arithmetic
2. **Live Second**: Validate integration
3. **E2E Last**: Validate complete system
4. **Automate Everything**: No manual regression testing

---

**Last Updated**: 2025-10-28
**Next Update**: After completing mock tests for all 13 rules
